File: app/src/test/java/com/leaguechampions/presenter/ChampionsPresenterTest.java
package com.leaguechampions.presenter;

import android.os.Bundle;
import android.view.MenuItem;

import com.leaguechampions.R;
import com.leaguechampions.datasource.remote.Api;
import com.leaguechampions.model.Champion;
import com.leaguechampions.model.RiotResponse;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.powermock.modules.junit4.PowerMockRunner;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@RunWith(PowerMockRunner.class)
public class ChampionsPresenterTest {

    private ChampionsPresenter presenter;

    private Map<String, Champion> data = new HashMap<String, Champion>() {{
        put("Riven", new Champion("Riven"));
        put("Jinx", new Champion("Jinx"));
    }};

    @Mock ChampionsPresenter.ChampionsViewable viewable;
    @Mock Api api;
    @Mock Call<RiotResponse> call;
    @Mock RiotResponse riotResponse;

    @Captor ArgumentCaptor<Callback<RiotResponse>> argumentCaptor;

    @Before
    public void setUp() {
        presenter = new ChampionsPresenter(api);
        presenter.setViewable(viewable);

        when(api.getChampions()).thenReturn(call);
        when(riotResponse.getData()).thenReturn(data);
    }

    @After
    public void tearDown() {

    }

    @Test
    public void testOnActivityCreated() {
        Bundle arguments = mock(Bundle.class);

        presenter.onActivityCreated(null, arguments);

        verify(api.getChampions()).enqueue(argumentCaptor.capture());
        argumentCaptor.getValue().onResponse(call,
                Response.success(riotResponse)
        );
        verify(viewable).setAdapter(riotResponse);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnActivityCreated_WhenResponseError400_ShowsError() throws Exception {
        Bundle arguments = mock(Bundle.class);

        presenter.onActivityCreated(null, arguments);

        verify(api.getChampions()).enqueue(argumentCaptor.capture());
        argumentCaptor.getValue().onResponse(call,
                Response.<RiotResponse>error(400, ResponseBody.create(MediaType.parse("application/json"), "{\"error\":\"failure\"}"))
        );
        verify(viewable).showError(R.string.error_code, 400);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnActivityCreated_WhenIOFailure_ShowsError() throws Exception {
        Bundle arguments = mock(Bundle.class);

        presenter.onActivityCreated(null, arguments);

        verify(api.getChampions()).enqueue(argumentCaptor.capture());
        argumentCaptor.getValue().onFailure(call,
                new IOException()
        );
        verify(viewable).showError(R.string.error_io);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnActivityCreated_WhenFailure_ShowsError() throws Exception {
        Bundle arguments = mock(Bundle.class);

        presenter.onActivityCreated(null, arguments);

        verify(api.getChampions()).enqueue(argumentCaptor.capture());
        argumentCaptor.getValue().onFailure(call,
                new Throwable()
        );
        verify(viewable).showError(R.string.error_something_went_wrong);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnOptionsItemSelected() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(R.id.menu_settings);

        boolean result = presenter.onOptionsItemSelected(item);

        assertThat(result).isEqualTo(true);
        verify(viewable).showSettings();
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnOptionsItemSelected_WhenItemNotHome_DoNothing() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(0);

        boolean result = presenter.onOptionsItemSelected(item);

        assertThat(result).isEqualTo(false);
        verifyNoMoreInteractions(viewable);
    }
}

File: features/champions/src/test/java/com/leaguechampions/features/champions/champions/ChampionsPresenterTest.kt
package com.leaguechampions.features.champions.champions

import org.junit.After
import org.junit.Before
import org.junit.Test

//@RunWith(PowerMockRunner::class)
//@PrepareForTest(RiotResponse::class, Champion::class)
class ChampionsPresenterTest {

//    private val data = hashMapOf(
//            "Riven" to Champion("", "Riven", "", ""),
//            "Jinx" to Champion("", "Jinx", "", "")
//    )
//
//    private lateinit var presenter: ChampionsPresenter
//
//    @Mock private lateinit var view: ChampionsPresenter.ChampionsView
//    @Mock private lateinit var api: Api
//    @Mock private lateinit var call: Call<RiotResponse>
//    @Mock private lateinit var riotResponse: RiotResponse
//    @Mock private lateinit var champion: Champion
//
//    @Captor lateinit var argumentCaptor: ArgumentCaptor<Callback<RiotResponse>>
//
    @Before
    fun setUp() {
//        presenter = ChampionsPresenter(view, api)
//
//        `when`(api.getChampions()).thenReturn(call)
//        `when`(riotResponse.data).thenReturn(data)
    }

    @After
    fun tearDown() {

    }

    @Test
    fun delete_me() {

    }

//    @Test
//    fun testOnActivityCreated() {
//        val arguments = mock(Bundle::class.java)
//
//        presenter.onActivityCreated(null, arguments)
//
//        verify(api.getChampions()).enqueue(argumentCaptor.capture())
//        argumentCaptor.value.onResponse(call,
//                Response.success(riotResponse)
//        )
//        verify(view).setAdapter(riotResponse)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnActivityCreated_WhenResponseError400_ShowsError() {
//        val arguments = mock(Bundle::class.java)
//
//        presenter.onActivityCreated(null, arguments)
//
//        verify(api.getChampions()).enqueue(argumentCaptor.capture())
//        argumentCaptor.value.onResponse(call,
//                Response.error(400, ResponseBody.create(MediaType.parse("application/json"), "{\"error\":\"failure\"}"))
//        )
//        verify(view).showError(R.string.error_code, 400)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnActivityCreated_WhenIOFailure_ShowsError() {
//        val arguments = mock(Bundle::class.java)
//
//        presenter.onActivityCreated(null, arguments)
//
//        verify(api.getChampions()).enqueue(argumentCaptor.capture())
//        argumentCaptor.value.onFailure(call,
//                IOException()
//        )
//        verify(view).showError(R.string.error_io)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnActivityCreated_WhenFailure_ShowsError() {
//        val arguments = mock(Bundle::class.java)
//
//        presenter.onActivityCreated(null, arguments)
//
//        verify(api.getChampions()).enqueue(argumentCaptor.capture())
//        argumentCaptor.value.onFailure(call,
//                Throwable()
//        )
//        verify(view).showError(R.string.error_something_went_wrong)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnOptionsItemSelected() {
//        val item = mock(MenuItem::class.java)
//        `when`(item.itemId).thenReturn(R.id.menu_settings)
//
//        val result = presenter.onOptionsItemSelected(item)
//
//        assertThat(result).isEqualTo(true)
//        verify(view).showSettings()
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnOptionsItemSelected_WhenItemNotHome_DoNothing() {
//        val item = mock(MenuItem::class.java)
//        `when`(item.itemId).thenReturn(0)
//
//        val result = presenter.onOptionsItemSelected(item)
//
//        assertThat(result).isEqualTo(false)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnItemClick() {
//        val version = "7.15.1"
//        val championId = "Riven"
//        `when`(champion.id).thenReturn(championId)
//
//        presenter.onItemClick(version, champion)
//
//        verify(view).showDetails(version, championId)
//        verifyNoMoreInteractions(view)
//    }
}
